iModelCore/ecobjects/src/ECSchema.cpp

```diff
@@ -3446,12 +3446,29 @@ POP_DISABLE_DEPRECATION_WARNINGS
     return status;
     }
 
+/*---------------------------------------------------------------------------------**//**
+* Returns true if thisSchema indirectly references potentiallyReferencedSchema
+* @bsimethod
++---------------+---------------+---------------+---------------+---------------+------*/
+static bool IndirectlyReferences(ECSchemaCR thisSchema, ECSchemaCR potentiallyReferencedSchema, ECN::SchemaMatchType matchType)
+    {
+    ECSchemaReferenceListCR referencedSchemas = thisSchema.GetReferencedSchemas();
+    if (referencedSchemas.end() != referencedSchemas.Find(potentiallyReferencedSchema.GetSchemaKey(), matchType))
+        return true;
+    for (ECSchemaReferenceList::const_iterator it = referencedSchemas.begin(); it != referencedSchemas.end(); ++it)
+        {
+        if (IndirectlyReferences(*it->second, potentiallyReferencedSchema, matchType))
+            return true;
+        }
+    return false;
+    }
+
 /*---------------------------------------------------------------------------------**//**
 * @bsimethod
 +---------------+---------------+---------------+---------------+---------------+------*/
 bool ECSchema::IsSchemaReferenced(ECSchemaCR thisSchema, ECSchemaCR potentiallyReferencedSchema)
     {
-    return ECSchema::IsSchemaReferenced(thisSchema, potentiallyReferencedSchema, ECN::SchemaMatchType::Exact);
+    return IndirectlyReferences(thisSchema, potentiallyReferencedSchema, ECN::SchemaMatchType::Exact);

```

Comment:
We should not change the behavior of IsSchemaReferenced.  It is really supposed to look in the direct references only.  If we were going to change the behavior of this overload of IsSchemaReferenced we would also have to change the other one so it was consistent.  Changing just one is very unexpected.  We should not change either.

The method 'DirectlyReferences' is confusingly named as it tests memory reference instead of name.  Luckily it is an internal method.

I kind of suspect that the root class passed into the convert method is not correct.  Does it only go one level deep?  You can always find the root class by walking the BaseProperty of each property until it's null ... that's the root class that defined the property first.

In general I think the algorithm is supposed to work at the base property then change derived properties as it goes.  In this case I guess it doesn't start at the base because the CA is defined on the leaf

When writing the test you should also validate that the intermediate property has been converted to the correct type.

Reply:
> I kind of suspect that the root class passed into the convert method is not correct. Does it only go one level deep?

The `ConvertToEnum()` method receives the root class `baseTest:TestClass` and traverses the class hierarchy recursively. However, when it attempts to set the type, the `PrimitiveECProperty::SetType()` method only checks if the `test:TestClass` class references `baseTest:TestClass`. This returns an `ECObjectsStatus::SchemaNotFound` status because we only check if `baseTest:TestClass` is directly referenced by `test:TestClass`.